### ISSUE TITLE
Installing NuGet packages is slow when project contains a csx file

### ISSUE THREAD

**Issue by reduckted** [daveaglick/Scripty#58]
After I've installed the Scripty.MsBuild package and added a csx file to the project, installing any other NuGet packages takes a _very_ long time, and locks up Visual Studio in the process.

Steps to reproduce:
1. Create a new project.
2. Run `Install-Package Scripty.MsBuild`.
3. Add a blank csx file to the project.
4. Run `Install-Package Scripy.Core`.

I've used the `Scripty.Core` package as an example here because it has a lot of dependencies which makes the slow-down very noticeable, but I've also had the same slow-down while installing xUnit.

As an example of the times I am seeing:

**Scripty.MsBuild NOT installed; With a csx file in the project**

```
PM> Install-Package Scripty.Core

Attempting to gather dependency information...
Gathering dependency information took 26.15 sec
Attempting to resolve dependencies for package 'Scripty.Core.0.5.0' with DependencyBehavior 'Lowest'
Resolving dependency information took 0 ms
...<snip...
Successfully installed 'Scripty.Core 0.5.0' to ConsoleApplication1
Executing nuget actions took 20.44 sec
Time Elapsed: 00:00:48.9877644
```

**Scripty.MsBuild installed; Without a csx file in the project**

```
PM> Install-Package Scripty.Core

Attempting to gather dependency information...
Gathering dependency information took 5.6 sec
Attempting to resolve dependencies for package 'Scripty.Core.0.5.0' with DependencyBehavior 'Lowest'
Resolving dependency information took 0 ms
...<snip>...
Successfully installed 'Scripty.Core 0.5.0' to ConsoleApplication1
Executing nuget actions took 18.45 sec
Time Elapsed: 00:00:25.4436322
```

**Scripty.MsBuild installed; With a csx file in the project**

```
PM> Install-Package Scripty.Core

Attempting to gather dependency information...
Gathering dependency information took 5.81 sec
Attempting to resolve dependencies for package 'Scripty.Core.0.5.0' with DependencyBehavior 'Lowest'
Resolving dependency information took 0 ms
...<snip>...
Successfully installed 'Scripty.Core 0.5.0' to ConsoleApplication1
Executing nuget actions took 3.39 min
Time Elapsed: 00:03:30.4221817
```

In summary, having a csx file in the project increased the install time by about three minutes.

**Comment by daveaglick** [comment]
Wow, crazy! I can't think of a reason why this might be happening, but there's a lot of complex ways NuGet interacts with the project. I'll look into it. What version of NuGet are you on?

**Comment by reduckted** [comment]
Visual Studio 2015 Update 3
NuGet v3.5.0.1484

I also have ReSharper 2016.2, but I disabled ReSharper and had the same result.

**Comment by reduckted** [comment]
So I've been running some experiments to see what might be causing the slow-down, and it appears that the task is running multiple times for each NuGet package that is installed.

I modified the `EvaluateScriptyFiles` task to look like this:

```
<Target Name="EvaluateScriptyFiles" DependsOnTargets="_ResolveScriptyFiles" BeforeTargets="BeforeCompile">
  <PropertyGroup>
    <Random>$([System.Guid]::NewGuid().ToString("n"))</Random>
    <OutputFilePath>D:\Temp\$(Random)</OutputFilePath>
  </PropertyGroup>

  <WriteLinesToFile File="$(OutputFilePath)" />
</Target>
```

When installing the `Scripty.Core` package, this created 164 files in that temp directory, meaning that the task ran 164 times! As a side-note, uninstalling a package does the same thing, but seems to run less often as there were only about 40 files produced.

The issue appears to be the target that `EvaluateScriptyFiles` depends on. I came across [this answer](http://stackoverflow.com/a/31081371/4397397) on StackOverflow which links to [this page](https://msdn.microsoft.com/en-us/library/ms171468%28v=vs.110%29.aspx). In short, it says this:

> Visual Studio attempts to execute targets with certain names when it loads a project. These targets include `Compile`, `ResolveAssemblyReferences`, `ResolveCOMReferences`, `GetFrameworkPaths`, and `CopyRunEnvironmentFiles`.

I suspect that `Compile` is being run after each NuGet package is installed (or run multiple times for each installed package), and that, of course, results in the `BeforeCompile` target being run, which causes `EvaluateScriptyFiles` to run.

The solution would be to depend on a different target.

You can see all of the targets that run as a result of the `Compile` target by calling MSBuild like this:

```
 msbuild /v:diag /t:compile
```

Compare that to the output for the `Build` target:

```
 msbuild /v:diag /t:build
```

For my project, the `Compile` target had 36 targets that ran, while the `Build` target had 68 targets that ran. Here's the targets that ran on `Build`, but _not_ on `Compile`:

```
_CheckForCompileOutputs
_CleanGetCurrentAndPriorFileWrites
_CopyAppConfigFile
_CopySourceItemsToOutputDirectory
_GenerateSatelliteAssemblyInputs
AfterBuild
AfterMarkupCompilePass1
AfterResGen
AssignTargetPaths
BeforeBuild
BeforeResGen
Build
BuildOnlySettings
CompileRdlFiles
CopyFilesToOutputDirectory
CoreBuild
CoreResGen
CreateCustomManifestResourceNames
CreateSatelliteAssemblies
ExpressionBuildExtension
FileClassification
GetCopyToOutputDirectoryItems
GetCopyToOutputDirectoryXamlAppDefs
GetTargetPath
IncrementalClean
PrepareForRun
PrepareRdlFiles
PrepareResourceNames
PrepareResources
ResGen
SplitResourcesByCulture
ValidationExtension
```

I tried changing the `EvaluateScriptyFiles`  target to use `BeforeTargets="BeforeBuild"`, and the slow-down went away. 

Running before `Build` and `CoreBuild` are probably other options too.

**Comment by daveaglick** [comment]
That's some amazing sleuth work! I'll try and carve out time this week to change the target and get a new release out. Thanks a ton for foguring out what was going on.

**Comment by daveaglick** [comment]
Obviously didn't get this done right away...been a busy month. It's a holiday weekend here, so time will be limited but I've set a reminder and am going to try and get this out in the next week or two. Thanks for your patience.
